[mlir python] Port in-tree dialects to nanobind.

hawkinsp opened the pull request:
This is a companion to #118583, although it can be landed independently because since #117922 dialects do not have to use the same Python binding framework as the Python core code.

This PR ports all of the in-tree dialect and pass extensions to nanobind, with the exception of those that remain for testing pybind11 support. It would make sense to merge this PR after merging #118583, if we have agreed that we are migrating the core to nanobind.

This PR also:
* removes CollectDiagnosticsToStringScope from NanobindAdaptors.h. This was overlooked in a previous PR and it is duplicated in Diagnostics.h.

hawkinsp committed:
[mlir python] Port in-tree dialects to nanobind.

This is a companion to #118583, although it can be landed independently
because since #117922 dialects do not have to use the same Python
binding framework as the Python core code.

This PR ports all of the in-tree dialect and pass extensions to nanobind,
with the exception of those that remain for testing pybind11 support. It
would make sense to merge this PR after merging #118583, if we have
agreed that we are migrating the core to nanobind.

This PR also:
* removes CollectDiagnosticsToStringScope from NanobindAdaptors.h. This
  was overlooked in a previous PR and it is duplicated in Diagnostics.h.

---------

Co-authored-by: Jacques Pienaar <jpienaar@google.com>

jpienaar reviewed (approved):
Overall looks good, I think we can do the following to avoid more of the nanobind warnings (else it just results in too much logging)

```diff
diff --git a/mlir/cmake/modules/AddMLIRPython.cmake b/mlir/cmake/modules/AddMLIRPython.cmake
index 67619a90c90b..e5952d1873ef 100644
--- a/mlir/cmake/modules/AddMLIRPython.cmake
+++ b/mlir/cmake/modules/AddMLIRPython.cmake
@@ -142,6 +142,9 @@ function(declare_mlir_python_extension name)
     mlir_python_DEPENDS ""
     mlir_python_BINDINGS_LIBRARY "${ARG_PYTHON_BINDINGS_LIBRARY}"
   )
+  if (LLVM_COMPILER_IS_GCC_COMPATIBLE OR CLANG_CL)
+    set_target_properties(${name} PROPERTIES INTERFACE_COMPILE_OPTIONS "-Wno-cast-qual;-Wno-zero-length-array;-Wno-extra-semi;-Wno-nested-anon-types;-Wno-pedantic")
+  endif()

   # Set the interface source and link_libs properties of the target
   # These properties support generator expressions and are automatically exported
diff --git a/mlir/python/CMakeLists.txt b/mlir/python/CMakeLists.txt
index 4f3a4e67d6f1..84b43a70ec61 100644
--- a/mlir/python/CMakeLists.txt
+++ b/mlir/python/CMakeLists.txt
@@ -476,9 +476,6 @@ declare_mlir_python_extension(MLIRPythonExtension.Core
     # Dialects
     MLIRCAPIFunc
 )
-if (LLVM_COMPILER_IS_GCC_COMPATIBLE OR CLANG_CL)
-set_target_properties(MLIRPythonExtension.Core PROPERTIES INTERFACE_COMPILE_OPTIONS "-Wno-cast-qual;-Wno-zero-length-array;-Wno-extra-semi;-Wno-nested-anon-types;-Wno-pedantic")
-endif()

 # This extension exposes an API to register all dialects, extensions, and passes
 # packaged in upstream MLIR and it is used for the upstream "mlir" Python
@@ -750,9 +747,6 @@ if(MLIR_INCLUDE_TESTS)
     EMBED_CAPI_LINK_LIBS
       MLIRCAPIPythonTestDialect
   )
-  if (LLVM_COMPILER_IS_GCC_COMPATIBLE OR CLANG_CL)
-    set_target_properties(MLIRPythonTestSources.PythonTestExtensionNanobind PROPERTIES INTERFACE_COMPILE_OPTIONS "-Wno-cast-qual;-Wno-zero-length-array;-Wno-extra-semi;-Wno-nested-anon-types;-Wno-pedantic")
-  endif()
 endif()
```

hawkinsp commented:
> Overall looks good, I think we can do the following to avoid more of the nanobind warnings (else it just results in too much logging)
...

Done

joker-eph commented:
> Overall looks good, I think we can do the following to avoid more of the nanobind warnings (else it just results in too much logging)

Instead of disabling warnings when building our sources, can we just disable them on nanobing headers? (that is defining the nanobind include path as system headers instead of regular include)

hawkinsp commented:
> > Overall looks good, I think we can do the following to avoid more of the nanobind warnings (else it just results in too much logging)
> 
> Instead of disabling warnings when building our sources, can we just disable them on nanobing headers? (that is defining the nanobind include path as system headers instead of regular include)

How does one do that, exactly?

makslevental commented:
> How does one do that, exactly?

https://gcc.gnu.org/onlinedocs/gcc/Diagnostic-Pragmas.html

Clang has the same thing but I can't find the docs right this moment

joker-eph commented:
In CMake it is the SYSTEM option of `include_directories`.
Since you're including nanobind using find_package, I think you don't have control on this. But you can remap it after find_package in the macro we use to inject the nanobind dependency like this:

```
# Retrieve the include directories set by nanobind
get_target_property(nanobind_includes nanobind::nanobind INTERFACE_INCLUDE_DIRECTORIES)
# Apply them as SYSTEM include directories
target_include_directories(my_target SYSTEM PRIVATE ${nanobind_includes})
```

hawkinsp commented:
> In CMake it is the SYSTEM option of `include_directories`. Since you're including nanobind using find_package, I think you don't have control on this. But you can remap it after find_package in the macro we use to inject the nanobind dependency like this:
> 
> ```
> # Retrieve the include directories set by nanobind
> get_target_property(nanobind_includes nanobind::nanobind INTERFACE_INCLUDE_DIRECTORIES)
> # Apply them as SYSTEM include directories
> target_include_directories(my_target SYSTEM PRIVATE ${nanobind_includes})
> ```

I spent 20 minutes trying to do this and failing. But I admit CMake is completely incomprehensible to me. Is this a hard blocker to landing it?

makslevental commented:
> I spent 20 minutes trying to do this and failing. But I admit CMake is completely incomprehensible to me. Is this a hard blocker to landing it?

Try with quotes like

`target_include_directories(my_target SYSTEM PRIVATE "${nanobind_includes}")`

CMake is very particular about what's a list/arg/string

jpienaar commented:
Its weird, with target\_include\_directories set, it affects the standalone example's python. It looks like the discovery of `_root_dir`.

makslevental commented:
> Its weird, with target_include_directories set, it affects the standalone example's python. It looks like the discovery of `_root_dir`.

It's because `INTERFACE` in `INTERFACE_INCLUDE_DIRECTORIES` means "propagate" (roughly I guess). 

So it occurs to me, do we actually need INTERFACE here

```
+  if (LLVM_COMPILER_IS_GCC_COMPATIBLE OR CLANG_CL)
+    set_target_properties(${name} PROPERTIES INTERFACE_COMPILE_OPTIONS "-Wno-cast-qual;-Wno-zero-length-array;-Wno-extra-semi;-Wno-nested-anon-types;-Wno-pedantic")
+  endif()
```

jpienaar commented:
I thought I removed those in the switch to system headers. But yes they were needed (at least cmake complained if I used private). But my cmake knowledge is of such a level that I should probably just ask Gemini for a solution :)

makslevental commented:
I got it - you need to fiddle with nanobind's library target itself https://github.com/wjakob/nanobind/blob/b7c4f1abfcab9cc5a8f0ef758926d92ff5eac3a3/cmake/nanobind-config.cmake#L358

which for us would look like 

```
target_compile_options(nanobind-static
  PRIVATE
  -Wno-cast-qual
  -Wno-zero-length-array
 ...
)
```

joker-eph commented on mlir/cmake/modules/AddMLIRPython.cmake:
```diff
@@ -661,6 +661,18 @@ function(add_mlir_python_extension libname extname)
+          -Wno-c++98-compat-extra-semi
+          -Wno-covered-switch-default
+      )
+    endif()
```
This can be removed now right?

joker-eph reviewed (approved):
LG!

jpienaar committed:
Move interop header include

Comment in Interop.h saying expectation on Windows is including this
after pybind11 - going to try the same for nanobind.

makslevental commented:
FYI after upgrading `mlir-python-extras` to latest I am suddenly getting this:

```
nanobind: leaked 3 instances!
 - leaked instance 0x7fe74e5c1c88 of type "mlir.execution_engine.ExecutionEngine"
 - leaked instance 0x7fe74e5ba868 of type "mlir._mlir_libs._mlir.ir.UnrankedMemRefType"
 - leaked instance 0x7fe74e67afa8 of type "mlir._mlir_libs.Context"
nanobind: leaked 126 types!
 - leaked type "mlir._mlir_libs._mlir.ir.AffineFloorDivExpr"
 - leaked type "mlir._mlir_libs._mlir.ir.AffineCeilDivExpr"
 - leaked type "mlir._mlir_libs._mlir.ir.F16Type"
 - leaked type "mlir._mlir_libs._mlir.ir.AffineMap"
 - leaked type "mlir._mlir_libs._mlirExecutionEngine.ExecutionEngine"
 - leaked type "mlir._mlir_libs._mlir.ir.FloatTF32Type"
 - leaked type "mlir._mlir_libs._mlir.ir.Float8E8M0FNUType"
 - leaked type "mlir._mlir_libs._mlir.ir.IntegerAttr"
 - leaked type "mlir._mlir_libs._mlir.ir.OpOperandIterator"
 - leaked type "mlir._mlir_libs._mlir.ir.OpView"
 - leaked type "mlir._mlir_libs._mlir.passmanager.PassManager"
 - ... skipped remainder
nanobind: leaked 860 functions!
 - leaked function "__init__"
 - leaked function ""
 - leaked function ""
 - leaked function ""
 - leaked function "dump"
 - leaked function "get_identity"
 - leaked function "__add__"
 - leaked function "maybe_downcast"
 - leaked function "__init__"
 - leaked function ""
 - leaked function ""
 - ... skipped remainder
nanobind: this is likely caused by a reference counting issue in the binding code.
```

I'm not sure if this is my fault or upstream's :shrug:

stellaraccident commented:
> FYI after upgrading `mlir-python-extras` to latest I am suddenly getting this:
> 
> ```
> nanobind: leaked 3 instances!
>  - leaked instance 0x7fe74e5c1c88 of type "mlir.execution_engine.ExecutionEngine"
>  - leaked instance 0x7fe74e5ba868 of type "mlir._mlir_libs._mlir.ir.UnrankedMemRefType"
>  - leaked instance 0x7fe74e67afa8 of type "mlir._mlir_libs.Context"
> nanobind: leaked 126 types!
>  - leaked type "mlir._mlir_libs._mlir.ir.AffineFloorDivExpr"
>  - leaked type "mlir._mlir_libs._mlir.ir.AffineCeilDivExpr"
>  - leaked type "mlir._mlir_libs._mlir.ir.F16Type"
>  - leaked type "mlir._mlir_libs._mlir.ir.AffineMap"
>  - leaked type "mlir._mlir_libs._mlirExecutionEngine.ExecutionEngine"
>  - leaked type "mlir._mlir_libs._mlir.ir.FloatTF32Type"
>  - leaked type "mlir._mlir_libs._mlir.ir.Float8E8M0FNUType"
>  - leaked type "mlir._mlir_libs._mlir.ir.IntegerAttr"
>  - leaked type "mlir._mlir_libs._mlir.ir.OpOperandIterator"
>  - leaked type "mlir._mlir_libs._mlir.ir.OpView"
>  - leaked type "mlir._mlir_libs._mlir.passmanager.PassManager"
>  - ... skipped remainder
> nanobind: leaked 860 functions!
>  - leaked function "__init__"
>  - leaked function ""
>  - leaked function ""
>  - leaked function ""
>  - leaked function "dump"
>  - leaked function "get_identity"
>  - leaked function "__add__"
>  - leaked function "maybe_downcast"
>  - leaked function "__init__"
>  - leaked function ""
>  - leaked function ""
>  - ... skipped remainder
> nanobind: this is likely caused by a reference counting issue in the binding code.
> ```
> 
> I'm not sure if this is my fault or upstream's 🤷

I've generally found the nanobind leak detector to make false positive warnings in a variety of situations that are common in these kind of programs. I can't tell if this is one of them. There is a function you can call at module init or after to silence it. I'd recommend landing that upstream while investigating. 

TBH - the only reliable way I've found to do leak detection is via a python debug build (which explicitly frees at shutdown) and ASAN. Anything else is easily thwarted with valid programs.
